File: jscode/src/modules/agent.ts
import { GetAgentStateResponse } from '@codebolt/types';
import cbws from './websocket';

const codeboltAgent = {
    /**
     * Retrieves an agent based on the specified task.
     * @param {string} task - The task for which an agent is needed.
     * @returns {Promise<AgentResponse>} A promise that resolves with the agent details.
     */
    getAgent: (task: string): Promise<any> => {
        return new Promise((resolve, reject) => {
            cbws.getWebsocket.send(JSON.stringify({
                "type": "agentEvent",
                "action": "getAgentByTask",
                "task": task
            }));
            cbws.getWebsocket.on('message', (data: string) => {
                const response = JSON.parse(data);
                if (response.type === "getAgentByTaskResponse") {
                    resolve(response); // Resolve the Promise with the agent details
                }
            });
        });
    },

    /**
     * Starts an agent for the specified task.
     * @param {string} task - The task for which the agent should be started.
     * @returns {Promise<void>} A promise that resolves when the agent has been successfully started.
     */
    startAgent: (agentId: string, task: string): Promise<any> => {
        return new Promise((resolve, reject) => {
            cbws.getWebsocket.send(JSON.stringify({
                "type": "agentEvent",
                "action": "startAgent",
                "agentId": agentId,
                "task": task
            }));
            cbws.getWebsocket.on('message', (data: string) => {
                const response = JSON.parse(data);
                if (response.type === "taskCompletionResponse") {
                    resolve(response); // Resolve the Promise when the agent has been successfully started
                }
            });
        });
    }
}


export default codeboltAgent;

File: modules/agent.py
import json
from modules.websocket import cbws_instance

class CodeboltAgent:
    def get_agent(self, task: str):
        """
        Retrieves an agent based on the specified task.
        :param task: The task for which an agent is needed.
        :return: A promise that resolves with the agent details.
        """
        async def get_agent_async():
            websocket = cbws_instance.get_websocket()
            await websocket.send(json.dumps({
                "type": "agentEvent",
                "action": "getAgentByTask",
                "task": task
            }))
            async for message in websocket:
                response = json.loads(message)
                if response.get("type") == "getAgentByTaskResponse":
                    return response

        return get_agent_async()

    def start_agent(self, agent_id: str, task: str):
        """
        Starts an agent for the specified task.
        :param agent_id: The ID of the agent to start.
        :param task: The task for which the agent should be started.
        :return: A promise that resolves when the agent has been successfully started.
        """
        async def start_agent_async():
            websocket = cbws_instance.get_websocket()
            await websocket.send(json.dumps({
                "type": "agentEvent",
                "action": "startAgent",
                "agentId": agent_id,
                "task": task
            }))
            async for message in websocket:
                response = json.loads(message)
                if response.get("type") == "taskCompletionResponse":
                    return response

        return start_agent_async()

codebolt_agent = CodeboltAgent()
